File: src/main/kotlin/com/github/hirsivaja/ip/packet/payload/udp/UdpHeader.kt
package com.github.hirsivaja.ip.packet.payload.udp

import java.nio.ByteBuffer

data class UdpHeader(val srcPort: UShort, val dstPort: UShort, val len: UShort, val checksum: Short) {

    fun encode(out: ByteBuffer) {
        out.putShort(srcPort.toShort())
        out.putShort(dstPort.toShort())
        out.putShort(len.toShort())
        out.putShort(checksum)
    }

    fun length() : Int {
        return UDP_HEADER_LEN
    }

    companion object {
        fun decode(buf: ByteBuffer) : UdpHeader {
            val srcPort = buf.getShort().toUShort()
            val dstPort = buf.getShort().toUShort()
            val len = buf.getShort().toUShort()
            val checksum = buf.getShort()
            return UdpHeader(srcPort, dstPort, len, checksum)
        }

        const val UDP_HEADER_LEN : Int = 8
    }
}

File: src/main/scala/com/github/hirsivaja/ip/payload/udp/UdpHeader.scala
package com.github.hirsivaja.ip.payload.udp

import java.nio.ByteBuffer

case class UdpHeader(srcPort: Short, dstPort: Short, len: Short, checksum: Short) {
  def this(srcPort: Short, dstPort: Short, len: Short) = {
    this(srcPort, dstPort, len, 0.toShort)
  }

  val length: Int = UdpHeader.UdpHeaderLen

  def encode(out: ByteBuffer): Unit = {
    out.putShort(srcPort)
    out.putShort(dstPort)
    out.putShort(len)
    out.putShort(checksum)
  }
}

object UdpHeader {
  val UdpHeaderLen: Int = 8

  def decode(in: ByteBuffer): UdpHeader = {
    val srcPort = in.getShort
    val dstPort = in.getShort
    val len = in.getShort
    val checksum = in.getShort
    new UdpHeader(srcPort, dstPort, len, checksum)
  }
}
